feat: add GitHub team allowlist configuration option

paulerickson opened the pull request:
## Summary
Configure permissions by GitHub team, with a list of teams and commands they are permitted to use.

## Usage
```sh
atlantis server --gh-team-allowlist "myteam:plan, secteam:apply"
```

## Background
This change is intended to replace #1206.
The code is the same, so all credit/blame goes to @jamengual & co-authors—I have only squashed those 75 revisions down to one, rebased on master, and resolved a few minor conflicts & incompatibilities.

msarvar commented:
@paulerickson hey, thanks for updating the PR. Original PR had quite a bit of comments seems like code stayed largely the same, would you be able to address the feedback from there? Comments are largely around moving the functionality into command runner and possibly moving the flag into server config.

majormoses commented:
I'd also like to ask that we consider the security implications of default allowing all teams to perform all actions. Maybe we should require people to set this and refuse any commands from users unless they are authorized? Regarding the moving of team allowlists around I think it should be controlled serverside in a config. I am wondering however if maybe it should even be configurable on a per repository basis. For example I might have a repository used to bootstrap new aws accounts into my organization and because of the required privileges this comes with we want it locked down to as few people as possible while others can be utilized by larger groups of people. Currently our threat model shows that the most critical points to cover currently due to lack of controls (which this helps build) is the repository itself and the verification of the payloads from Github webhook.

paulerickson commented:
I have moved the permissions check from event handler to command runner, which I think someone requested: b8fce3a

msarvar commented:
@paulerickson Hey, pr looks good. Left couple comments, let me know what you think

paulerickson commented on server/events/vcs/github_client.go:
```diff
@@ -386,6 +386,33 @@ func (g *GithubClient) MarkdownPullLink(pull models.PullRequest) (string, error)
+	for {
+		teams, resp, err := g.client.Teams.ListTeams(g.ctx, org, opts)
+		if err != nil {
+			return nil, err
```
addressed in https://github.com/runatlantis/atlantis/pull/1694/commits/ac049eaade6232e6a6dde0ff8dea3a37e5fcdd8c

paulerickson commented on server/events/vcs/github_client.go:
```diff
@@ -386,6 +386,33 @@ func (g *GithubClient) MarkdownPullLink(pull models.PullRequest) (string, error)
+		}
+		for _, t := range teams {
+			membership, _, err := g.client.Teams.GetTeamMembershipBySlug(g.ctx, org, *t.Slug, user.Username)
+			if err == nil && membership != nil {
```
I logged it here: https://github.com/runatlantis/atlantis/pull/1694/commits/ac049eaade6232e6a6dde0ff8dea3a37e5fcdd8c

I think continuing with no groups is a simple way to handle it unless we can retry or do some better handling in a higher layer

paulerickson commented on server/events/team_allowlist_checker_test.go:
```diff
@@ -0,0 +1,35 @@
+	Equals(t, true, checker.IsCommandAllowedForAnyTeam(teams, `plan`))
+	Equals(t, true, checker.IsCommandAllowedForAnyTeam(teams, `release`))
+	Equals(t, false, checker.IsCommandAllowedForAnyTeam(teams, `noop`))
+}
```
Added in https://github.com/runatlantis/atlantis/pull/1694/commits/23c5cd9833a2e0d76983249961abebeda39b54a7

paulerickson commented on server/events/team_allowlist_checker.go:
```diff
@@ -0,0 +1,72 @@
+// IsCommandAllowedForAnyTeam returns true if any of the teams is allowed to execute the command
+// and false otherwise.
+func (checker *TeamAllowlistChecker) IsCommandAllowedForAnyTeam(teams []string, command string) bool {
+	c := strings.TrimSpace(command)
```
https://github.com/runatlantis/atlantis/pull/1694/commits/fbcd74f7d1d817f484f67cfb08aed842279219dc

paulerickson commented on server/events/team_allowlist_checker.go:
```diff
@@ -0,0 +1,72 @@
+// and false otherwise.
+func (checker *TeamAllowlistChecker) IsCommandAllowedForTeam(team string, command string) bool {
+	t := strings.TrimSpace(team)
+	c := strings.TrimSpace(command)
```
Not that I can see.  Removed in https://github.com/runatlantis/atlantis/pull/1694/commits/fbcd74f7d1d817f484f67cfb08aed842279219dc

paulerickson commented:
> @paulerickson Hey, pr looks good. Left couple comments, let me know what you think

Thanks, @msarvar.  I think I've addressed everything, and rebased from master

msarvar reviewed (approved):
Looks good, thanks for contributing!

osterman commented:
Relates to https://github.com/runatlantis/atlantis/issues/308

albertorm95 commented on server/events/vcs/github_client.go:
```diff
@@ -454,6 +454,35 @@ func (g *GithubClient) MarkdownPullLink(pull models.PullRequest) (string, error)
+			return nil, errors.Wrap(err, "retrieving GitHub teams")
+		}
+		for _, t := range teams {
+			membership, _, err := g.client.Teams.GetTeamMembershipBySlug(g.ctx, org, *t.Slug, user.Username)
```
What about the rate limiting? Wound't it be better to use the GraphQL so we could just pull the orgs teams of a user?

https://docs.github.com/en/rest/overview/resources-in-the-rest-api

paulerickson replied:
:+1: I haven't worked with the GraphQL API, but it does seem like some calls could be saved that way

albertorm95 replied:
@paulerickson Thanks for the answer I could try to switch this feature to use GraphQL :)

nitrocode commented:
@marcportabellaclotet-mt please open a new issue for that. Sounds like a good idea.

marcportabellaclotet-mt commented:
@nitrocode  [done](https://github.com/runatlantis/atlantis/issues/2726)
